### PR TITLE
Jinja2Bear: Add support for whitespace control

### PR DESCRIPTION
**Pull request by kriti21**
This adds support for whitespace control for
statement tags and when trim_blocks are used
with statement or control tags.

Closes https://github.com/coala/coala-bears/issues/2002

### CODE REVIEW

File: tests/jinja2/Jinja2BearTest.py [hunk 1]
```diff
@@ -23,6 +23,25 @@
...
+
+Jinja2BearCustomStatementSpacingTest = verify_local_bear(
+    Jinja2Bear,
+    valid_files=(r'{%statement1%}',
+                 r'{%statement2 foobar%}'),
+    invalid_files=(r'{%foo bar %}',
+                   r'{% foo bar%}'
+                   r'{% foo bar %}'
+                   r'{%   foo bar  %}'),
+    settings={'statement_spacing': '0'})
```
Comment (khanchi97) [inline tests/jinja2/Jinja2BearTest.py]: I think you can put this zero without single quotes.

Comment (kriti21) [reply tests/jinja2/Jinja2BearTest.py]: @khanchi97 That's true but I don't think it creates much of a difference here. Other similar tests have 0 within single quotes.

File: bears/jinja2/Jinja2Bear.py [hunk 2]
```diff
@@ -184,6 +186,45 @@ def check_for_variable_spacing_issues(self,
                     end_column=m.end(0) + 1,
                     diffs=diff)
 
+    def check_for_statement_spacing_issues(self,
+                                           file,
+                                           filename,
+                                           line,
+                                           line_number,
+                                           statement_spacing):
```
Comment (newbazz) [inline bears/jinja2/Jinja2Bear.py]: You have to give default value here right? and datatype isnt defined as well ;)

Comment (kriti21) [reply bears/jinja2/Jinja2Bear.py]: Default values and their data type has been defined in the run function already.

**Comment by kriti21** [comment]
@Makman2 your guidance really made my day. This one finally passed all the tests. Please review :)

File: bears/jinja2/Jinja2Bear.py [hunk 3]
```diff
@@ -112,11 +112,16 @@ class Jinja2Bear(LocalBear):
 
     VARIABLE_REGEX = re.compile(
         r'(?P<open>{{)(?P<content>.*?)(?P<close>}})')
+    STATEMENT_REGEX = re.compile(
+        r'(?P<open>{%[+-]?)'
+        # Ignore PycodestyleBear (E501) LineLengthBear
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: nah nah, please break the string properly ;)

File: bears/jinja2/Jinja2Bear.py [hunk 4]
```diff
@@ -294,7 +338,6 @@ def check_for_empty_control_stack(self, filename):
         """
         Checks if the control stack is empty and yields a Result for each
         item left on the stack.
-
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: unnecessary change

File: bears/jinja2/Jinja2Bear.py [hunk 5]
```diff
@@ -321,6 +365,11 @@ def run(self,
             this: ``{{ var_name }}``. This can be set to any number of spaces
             via the setting ``variable_spacing``.
             Malformatted variable tags are detected and fixes suggested.
+        * Statement spacing:
+            Statement tags should be padded with one space on each side, like
+            this: ``{% statement %}``. This can be set to any number of spaces
+            via the setting ``statement_spacing``.
+            Malformatted statement tags are detected and fixes suggested.
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: I think you can leave this sentence, it's repeated from before. I think users understand that coala could propose changes :)

File: bears/jinja2/Jinja2Bear.py [hunk 6]
```diff
@@ -345,6 +394,9 @@ def run(self,
         :param variable_spacing:
             The number of spaces a variable block should be spaced with.
             Default is 1.
+        :param statement_spacing:
+            The number of spaces a statement block should be spaced with.
+            Default is 1.
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: hm it's actually redundant to mention the default value, coala displays them automatically when querying bear infos. But due to consistency you can keep it. We file an issue after this PR to remove this.

File: bears/jinja2/Jinja2Bear.py [hunk 7]
```diff
@@ -112,11 +112,16 @@ class Jinja2Bear(LocalBear):
 
     VARIABLE_REGEX = re.compile(
         r'(?P<open>{{)(?P<content>.*?)(?P<close>}})')
+    STATEMENT_REGEX = re.compile(
+        r'(?P<open>{%[+-]?)'
+        # Ignore PycodestyleBear (E501) LineLengthBear
+        r'(?P<content>(?![+-]?\s*for)(?![+-]?\s*if)(?![+-]?\s*endfor)(?![+-]?\s*endif).*?)'
+        r'(?P<close>[+-]?%})')
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: Isn't this covered by `CONTROL_START_REGEX` and `CONTROL_END_REGEX`?

File: tests/jinja2/Jinja2BearTest.py [hunk 8]
```diff
@@ -39,6 +58,16 @@ def test_variable_spacing(self):
...
 
+    def test_statement_spacing(self):
+        content = (r'foo {%-statement1 %} bar',)
+        with execute_bear(self.uut, 'F', content) as result:
+            self.assertEqual(result[0].diffs['F'].unified_diff,
+                             '--- \n'
+                             '+++ \n'
+                             '@@ -1 +1 @@\n'
+                             '-foo {%-statement1 %} bar\n'
+                             '+foo {%- statement1 %} bar')
```
Comment (Makman2) [inline tests/jinja2/Jinja2BearTest.py]: could you also test with `{%-statement1%}` and `{%- statement1%}` and `{%-     statement1   %}` ? :)

Comment (Makman2) [reply tests/jinja2/Jinja2BearTest.py]: In fact you can move your `verify_local_bear` tests here ;)

Comment (Makman2) [reply tests/jinja2/Jinja2BearTest.py]: though... actually nvm, I think your tests using `verify_local_bear` are actually quite good.

File: bears/jinja2/Jinja2Bear.py [hunk 9]
```diff
@@ -184,6 +189,45 @@ def check_for_variable_spacing_issues(self,
...
+                yield Result.from_values(
+                    origin=self,
+                    message='Statement blocks should be spaced with '
+                            '`{0}` spaces on each side.'.format(
+                                statement_spacing),
+                    file=filename,
+                    line=line_number,
+                    column=m.start(0) + 1,
+                    end_line=line_number,
+                    end_column=m.end(0) + 1,
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: you don't have to specify `0` for `start()` and `end()`, it's default.

File: bears/jinja2/Jinja2Bear.py [hunk 10]
```diff
@@ -184,6 +189,45 @@ def check_for_variable_spacing_issues(self,
...
+        for m in self.STATEMENT_REGEX.finditer(line):
+            match = m.group('content')
+            if not has_required_spacing(match, statement_spacing):
+                diff = generate_spacing_diff(
+                    file, filename, line, line_number, m,
+                    statement_spacing)
+                yield Result.from_values(
+                    origin=self,
+                    message='Statement blocks should be spaced with '
+                            '`{0}` spaces on each side.'.format(
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: unnecessary `0` inside format placeholder :)

Comment (Makman2) [reply bears/jinja2/Jinja2Bear.py]: I would rephrase the message a bit:
```
Statement blocks not spaces with {} spaces on each side.
```

Comment (Makman2) [reply bears/jinja2/Jinja2Bear.py]: So it gets a more passive voice. Using `should` could sound like some recommendation from our side, although the coala-user was the one specifying the option ;)

**Comment by Makman2** [comment]
@kriti21 what about
> - End tag label does not match expected label
> - Control start tag has no corresponding end

Is this included? If not I would anyway file new issues for them, they can be easily separated.

**Comment by Makman2** [comment]
unack 2834786b4b1bad86866d42cdf50c9dbb96c43c1c

File: bears/jinja2/Jinja2Bear.py (hunk 3 above)
Comment (kriti21) [reply bears/jinja2/Jinja2Bear.py]: @Makman2 Please provide some more detail on how exactly should i improve this. I am not able to get it clearly.

File: bears/jinja2/Jinja2Bear.py (hunk 6 above)
Comment (kriti21) [reply bears/jinja2/Jinja2Bear.py]: Will keep in mind to open an issue for this once this PR approved. :)

File: bears/jinja2/Jinja2Bear.py (hunk 7 above)
Comment (kriti21) [reply bears/jinja2/Jinja2Bear.py]: No this was not covered b `CONTROL_START_REGEX` and `CONTROL_END_REGEX`. In fact if I do not include this, all my tests for statement blocks would fail.

File: tests/jinja2/Jinja2BearTest.py (hunk 8 above)
Comment (kriti21) [reply tests/jinja2/Jinja2BearTest.py]: >could you also test with {%-statement1%} and {%- statement1%} and {%- statement1 %}


Done !
> In fact you can move your verify_local_bear tests here

Do i need to move the tests in `verify_local_bear` also?

**Comment by kriti21** [comment]
>- End tag label does not match expected label
>- Control start tag has no corresponding end

These are already detected by `Jinja2Bear` . The only issue that is still to be addressed is that some tests for control start tags also apply for statement tags  (if i write a more general regex). So perhaps we could file two issues once this is approved.

1. To correct mistakes in the control start tags and
2. And then improve regex for statement tags as well.

File: bears/jinja2/Jinja2Bear.py (hunk 3 above)
Comment (Makman2) [reply bears/jinja2/Jinja2Bear.py]: -->
```
(r'string-that-is-ver'
 r'y-long')
```
:)

File: tests/jinja2/Jinja2BearTest.py (hunk 8 above)
Comment (Makman2) [reply tests/jinja2/Jinja2BearTest.py]: no really keep them as they are 😅 (I should have removed my comments 😅 )

File: bears/jinja2/Jinja2Bear.py (hunk 7 above)
Comment (Makman2) [reply bears/jinja2/Jinja2Bear.py]: Can you reuse the `CONTROL_..._REGEX` variables in your `STATEMENT_REGEX`?

File: tests/jinja2/Jinja2BearTest.py (hunk 8 above)
Comment (kriti21) [reply tests/jinja2/Jinja2BearTest.py]: Okay Got it :)

File: bears/jinja2/Jinja2Bear.py [hunk 11]
```diff
@@ -112,11 +112,17 @@ class Jinja2Bear(LocalBear):
 
     VARIABLE_REGEX = re.compile(
         r'(?P<open>{{)(?P<content>.*?)(?P<close>}})')
+    STATEMENT_REGEX = re.compile(
+        r'(?P<open>{%[+-]?)'
+        r'(?P<content>(?![+-]?\s*for)'
+        r'(?![+-]?\s*if)(?![+-]?\s*endfor)'
+        r'(?![+-]?\s*endif).*?)'
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: hm can you maybe simplify this to `\s(?!if).*?` or so? You repeat `[+-]` very often and it's already being matched in `<open>`

File: bears/jinja2/Jinja2Bear.py [hunk 12]
```diff
@@ -184,6 +190,45 @@ def check_for_variable_spacing_issues(self,
...
 
+    def check_for_statement_spacing_issues(self,
+                                           file,
+                                           filename,
+                                           line,
+                                           line_number,
+                                           statement_spacing):
+        """
+        Checks any statement in the given line for spacing issues.
+        Yields a Result for each issue found.
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: You can remove that sentence ;) That's obvious :)

File: bears/jinja2/Jinja2Bear.py [hunk 13]
```diff
@@ -184,6 +190,45 @@ def check_for_variable_spacing_issues(self,
...
+            The name of the file currently being inspected.
+        :param line:
+            The content of the line currently being inspected.
+        :param line_number:
+            The current line number.
+        :param statement_spacing:
+            The number of spaces required on each side of a statement tag.
+        """
+        for m in self.STATEMENT_REGEX.finditer(line):
+            match = m.group('content')
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: optional: Maybe call `m` --> `match`, and `match` --> `content`

**Comment by Makman2** [comment]
unack c0c86ab1917539ec8c2671e5faec6fe64e336a74

File: bears/jinja2/Jinja2Bear.py [hunk 14]
```diff
@@ -112,11 +112,17 @@ class Jinja2Bear(LocalBear):
 
     VARIABLE_REGEX = re.compile(
         r'(?P<open>{{)(?P<content>.*?)(?P<close>}})')
+    STATEMENT_REGEX = re.compile(
+        r'(?P<open>{%[+-]?)'
+        r'(?P<content>(?!for)'
+        r'(?!if)(?!endfor)'
+        r'(?!endif).*?)'
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: I think it would be good for the sake of readability to do it like
```python
r'(?P<open>{%[+-]?)'
r'(?P<content>'
r'(?!for)(?!if)(?!endfor)(?!endif)'
r'.*?)'
```
Is it also possible to use the `|` operator?
```
(?!for|if|endfor|endif)
```

File: bears/jinja2/Jinja2Bear.py [hunk 15]
```diff
@@ -184,6 +190,44 @@ def check_for_variable_spacing_issues(self,
...
+        """
+        for match in self.STATEMENT_REGEX.finditer(line):
+            content = match.group('content')
+            if not has_required_spacing(content, statement_spacing):
+                diff = generate_spacing_diff(
+                    file, filename, line, line_number, match,
+                    statement_spacing)
+                yield Result.from_values(
+                    origin=self,
+                    message='Statement blocks not spaces with '
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: 1. block~~s~~
2. space~~s~~**d**
3. Why "blocks" actually? Not just "Statement not spaced with ... spaces"?

File: bears/jinja2/Jinja2Bear.py (hunk 14 above)
Comment (kriti21) [reply bears/jinja2/Jinja2Bear.py]: Nope this did not work. I tried it.

File: bears/jinja2/Jinja2Bear.py [hunk 16]
```diff
@@ -184,6 +190,44 @@ def check_for_variable_spacing_issues(self,
...
+                            '`{}` spaces on each side.'.format(
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: Hm it's a number, I think you shouldn't enclose it with backticks inside the message, I mean it appear in the console :)

File: bears/jinja2/Jinja2Bear.py (hunk 14 above)
Comment (kriti21) [reply bears/jinja2/Jinja2Bear.py]: `!` needs to applied for each one separately.

File: bears/jinja2/Jinja2Bear.py [hunk 17]
```diff
@@ -184,6 +190,44 @@ def check_for_variable_spacing_issues(self,
...
+                                statement_spacing),
+                    file=filename,
+                    line=line_number,
+                    column=match.start() + 1,
+                    end_line=line_number,
+                    end_column=match.end() + 1,
+                    diffs=diff)
```
Comment (Makman2) [inline bears/jinja2/Jinja2Bear.py]: >         :param diffs:
>            A dictionary with filename as key and ``Diff`` object
>            associated with it as value.

Comment (Makman2) [reply bears/jinja2/Jinja2Bear.py]: Nvm, the variable name is just a bit misleading :) But it's okay, as the existing code calls it the same.

**Comment by Makman2** [comment]
unack ca364f7cedbb5d62e5c15906042294d8a5a4151f

File: bears/jinja2/Jinja2Bear.py (hunk 17 above)
Comment (kriti21) [reply bears/jinja2/Jinja2Bear.py]: `diff` is not a parameter for the function I have written here. It is being passed to Result class. Should i still define this as well?

Comment (Makman2) [reply bears/jinja2/Jinja2Bear.py]: No no keep it as is, `generate_spacing_diff` wraps everything properly inside a dict like required :)

File: tests/jinja2/Jinja2BearTest.py [hunk 18]
```diff
@@ -88,6 +133,20 @@ def test_control_spacing(self):
...
 good_file3 = '{% for x in y %} one liner needs no label {% endfor %}'
 
+good_file4 = """
+foo
+{%+ for x in something -%}
+render stuff
+{%+ endfor -%}{# for x in something #}
+"""
+
+good_file4 = """
```
Comment (Makman2) [inline tests/jinja2/Jinja2BearTest.py]: --> `good_file5` ;)

File: tests/jinja2/Jinja2BearTest.py [hunk 19]
```diff
@@ -88,6 +133,20 @@ def test_control_spacing(self):
...
+foo
+{%- for x in something +%}
+render stuff
+{%- endfor +%}{# for x in something #}
+"""
```
Comment (Makman2) [inline tests/jinja2/Jinja2BearTest.py]: However, these are all unused variables, aren't they?

**Comment by Makman2** [comment]
Now this is an extensive test set ;) However some final polishing needed :P

**Comment by Makman2** [comment]
unack 99c8da0409a41c81c4273adaeaec174dede816bf

**Comment by kriti21** [comment]
@Makman2 The tests i added yesterday were not completely correct and so the simplified regex gave no error. As you can see, the tests now are better and more correct but this is the only regex with which `pytest` runs successfully. I tried a lot to simplify it, but even removing minor things like `[+-]` cause `pytest` to fail. I do have certain proposals to improve control_start regex like - 
1. control_start_regex should start with `for` or `if` or `endfor` or `endif` and therefore,
2. control_start tags test should be applied to tags beginning with `for` and `if` which is not the case right now. Also another option is to segregate handling tags and white spaces for control tags which is (to some extent) contained in one function only. (This might make the process slightly redundant though)
I want to open these issue ones this PR is approved. Review this one please :)

**Comment by Makman2** [comment]
ack c1fb2f47c694292e7a44257b535b86a8e30300db

**Comment by Makman2** [comment]
@gitmate-bot rebase

**Comment by Makman2** [comment]
Marking approved, unrelated CI fail.

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
@gitmate-bot ff
